csrc/id_model/id_model.cpp

```diff
@@ -837,4 +845,266 @@ ValGraph IdModel::buildIntersection(
   return intersection;
 }
 
+namespace {
+
+// Check if there's an equivalent expression as iel_expr that uses
+// maybe_promoted_inputs. This is used to avoid redundantly replaying
+// expressions.
+// NOTE: This is currently overly conservative and some
+// opportunities for reuse are lost, althought it doesn't affect
+// the correctness of the analysis.
+Expr* findMatchingExpr(
+    const ExprGroup& iel_expr,
+    const ValGraph& iel_graph,
+    const std::vector<IterDomain*>& maybe_promoted_inputs) {
+  // Grab all uses of the promoted inputs
+  ExprGroups maybe_promoted_input_uses;
+  for (auto inp_id : maybe_promoted_inputs) {
+    // inp_id may have been just replayed, in which case it should
+    // not exist in the IEL graph. It should be just ignored as it
+    // should not have any use yet.
+    if (!iel_graph.hasGroup(inp_id)) {
+      continue;
+    }
```

Comment:
Is it possible that all ids in `maybe_promoted_inputs` are just replayed?

Reply:
Yes. In that case, there's no existing use expr, so just `nullptr` would be returned, meaning no reuse would happen and a new expr would be replayed.